python/tank_vendor/shotgun_deploy/descriptor_io/factory.py

```diff
@@ -0,0 +1,170 @@
+# Copyright (c) 2016 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import urllib
+from ..errors import ShotgunDeployError
+from .. import constants
+from .. import util
+log = util.get_shotgun_deploy_logger()
+
+def create_io_descriptor(sg, descriptor_type, location, bundle_cache_root, fallback_roots):
+    """
+    Factory method. Use this method to construct all DescriptorIO instances.
+
+    :param sg: Shotgun connection to associated site
+    :param descriptor_type: Either AppDescriptor.APP, CORE, ENGINE or FRAMEWORK
+    :param location: A std location dictionary dictionary or string
+    :param bundle_cache_root: Root path to where downloaded apps are cached
+    :param fallback_roots: List of immutable fallback cache locations where
+                           apps will be searched for. Note that when descriptors
+                           download new content, it always ends up in the
+                           bundle_cache_root.
+    :returns: Descriptor object
+    """
+    from .appstore import IODescriptorAppStore
+    from .dev import IODescriptorDev
+    from .path import IODescriptorPath
+    from .shotgun_entity import IODescriptorShotgunEntity
+    from .git import IODescriptorGit
+    from .git_branch import IODescriptorGitBranch
+    from .manual import IODescriptorManual
+
+    if isinstance(location, basestring):
+        # translate uri to dict
+        location_dict = location_uri_to_dict(location)
+    else:
+        location_dict = location
+
+    if location_dict.get("type") == "app_store":
```

Comment:
Let's do a meta-baseclass that automatically registers the derived class automatically when you derive from it! :)